src/coreclr/System.Private.CoreLib/src/System/Runtime/CompilerServices/RuntimeHelpers.CoreCLR.cs

```diff
@@ -637,6 +654,99 @@ public bool IsTypeDesc
 
             return (MethodTable*)m_asTAddr;
         }
+
+        /// <summary>
+        /// Gets a <see cref="MethodTable"/> pointer for the current type handle.
+        /// </summary>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public MethodTable* GetMethodTable()
+        {
+            if (IsTypeDesc)
+            {
+                return AsTypeDesc()->GetMethodTable();
+            }
+
+            return (MethodTable*)m_asTAddr;
+        }
+    }
+
+    internal unsafe struct TypeDesc
+    {
+        /// <summary>
+        /// <para>
+        /// The low-order 8 bits of this flag are used to store the <see cref="CorElementType"/>,
+        /// which discriminates what kind of <see cref="TypeDesc"/> this is.
+        /// </para>
+        /// <para>The remaining bits are available for flags.</para>
+        /// </summary>
+        private readonly uint TypeAndFlags;
+
+        /// <summary>
+        /// Gets a <see cref="MethodTable"/> pointer for the current type desc.
+        /// </summary>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public MethodTable* GetMethodTable()
+        {
+            // The native version here has a check for IsGenericVariable, which causes this method to return null.
+            // This path is never taken by TypedReference, as it always assumes the returned MethodTable* pointer
+            // is not null. If adding code that needs this path to be available as well, it needs to be ported too.
+            //
+            // if (IsGenericVariable)
+            // {
+            //     return null;
+            // }
+
+            if (GetInternalCorElementType() == CorElementType.ELEMENT_TYPE_FNPTR)
+            {
+                return RuntimeTypeHandle.GetElementTypeMethodTable(CorElementType.ELEMENT_TYPE_U);
```

Comment:
You should not need FCall for this. The JIT should be able to inline the MethodTable here. 

For NativeAOT, we have `MethodTableOf<T>`. Can we enable it for regular CoreCLR too?

Reply:
`RuntimeTypeHandle.GetValueInternal(typeof(UIntPtr).Handle)` may also do the job and it should not require any JIT/VM interface changes.

Reply:
> "You should not need FCall for this. The JIT should be able to inline the MethodTable here."

The reason I introduced that FCall was that I wanted to avoid having to duplicate so much VM code in C# just to handle that (presumably niche) path here, and I couldn't find another way to do the same logic in managed. I had not noticed that `RuntimeTypeHandle.GetValueInternal(typeof(UIntPtr).TypeHandle)` could be used here, does that mean that that's effectively the equivalent of `MethodTableOf<T>` for CoreCLR? Should we document that API as such to help others? Right now what that `RuntimeTypeHandle.GetValueInternal` does exactly is not really that intuitive, even looking at the FCall impl (at least for me, it wasn't immediately obvious it was returning a `MethodTable*` for the input type handle). Should we also change the return type to just be `MethodTable*` to be more explicit, if that's what it does anyway? Right now it's just `IntPtr` 🙂

Reply:
> it wasn't immediately obvious it was returning a MethodTable* for the input type handle

It returns MethodTable or TypeDesc, depending on what you pass in.